packages/flutter_tools/lib/src/ios/fallback_discovery.dart

```diff
@@ -0,0 +1,156 @@
+    @required String packageName,
+  }) async {
+    final Uri result = await _attemptServiceConnection(
+      assumedDevicePort: assumedDevicePort,
+      hostVmservicePort: hostVmservicePort,
+      packageName: packageName,
+    );
+    if (result != null) {
+      return result;
+    }
+
+    try {
+      final Uri result = await _mDnsObservatoryDiscovery.getObservatoryUri(
+        packageId,
+        deivce,
+        usesIpv6: usesIpv6,
+        hostVmservicePort: hostVmservicePort,
+      );
+      if (result != null) {
+        UsageEvent('ios-mdns', 'success').send();
+        return result;
+      }
+    } on Exception catch (err) {
+      _logger.printTrace(err.toString());
+    }
+    _logger.printTrace('Failed to connect with mDNS, falling back to log scanning');
+    UsageEvent('ios-mdns', 'failure').send();
+
+    try {
+      final Uri result = await _protocolDiscovery.uri;
+      UsageEvent('ios-mdns', 'fallback-success').send();
+      return result;
+    } on ArgumentError {
+    // In the event of an invalid InternetAddress, this code attempts to catch
+    // an ArgumentError from protocol_discovery.dart
+    } on Exception catch (err) {
+      _logger.printTrace(err.toString());
+    }
+    _logger.printTrace('Failed to connect with log scanning');
+    UsageEvent('ios-mdns', 'fallback-failure').send();
+    return null;
+  }
+
+  // Attempt to connect to the VM service and find an isolate with a matching `packageName`.
+  // Returns `null` if no connection can be made.
+  Future<Uri> _attemptServiceConnection({
+    @required int assumedDevicePort,
+    @required int  hostVmservicePort,
+    @required String packageName,
+  }) async {
+    int hostPort;
+    Uri assumedWsUri;
+    try {
+      hostPort = await _portForwarder.forward(assumedDevicePort, hostPort: hostVmservicePort);
+      assumedWsUri = Uri.parse('ws://localhost:$hostPort/ws');
+    } on Exception catch (err) {
+      _logger.printTrace(err.toString());
+      _logger.printTrace('Failed to connect directly, falling back to mDNS');
+      UsageEvent('ios-handshake', 'failure').send();
+      return null;
+    }
+
+    // Attempt to connect to the VM service 5 times.
+    int attempts = 0;
+    const int kDelaySeconds = 2;
+    while (attempts < 5) {
+      try {
+        final VmService vmService = await _vmServiceConnectUri(assumedWsUri.toString());
+        final VM vm = await vmService.getVM();
+        for (final IsolateRef isolateRefs in vm.isolates) {
+          final Isolate isolate = await vmService.getIsolate(isolateRefs.id) as Isolate;
+          final LibraryRef library = isolate.rootLib;
+          if (library.uri.startsWith('package:$packageName')) {
+            UsageEvent('ios-handshake', 'success').send();
+            return Uri.parse('http://localhost:$hostPort');
+          }
+        }
+      } on Exception {
+        // No action, we might have failed to connect.
+        // The error message will be irrelevant.
```

Comment:
You might record one of the exception objects to print with the failure trace message.

Reply:
Done